CDAP-4075 Added lifecycle methods for the Workflow.

sagarkapare opened the pull request:
JIRA: https://issues.cask.co/browse/CDAP-4075

This PR adds `initialize` and `destroy` methods to the Workflow. 
Next PR will contain following changes - 
1. Ability to get the map of node ids to the `WorkflowNodeState` from `WorkflowContext`. The dependent PR is already opened here #5326 
2. Ability to get the status of the Workflow through `WorkflowContext`.

sagarkapare commented:
Build is running here - http://builds.cask.co/browse/CDAP-DUT3732-1

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/AbstractWorkflow.java:
```diff
@@ -189,4 +190,21 @@ protected final void createLocalDataset(String datasetName, Class<? extends Data
+  /**
+   * Return an instance of the {@link WorkflowContext} when the Workflow is running, otherwise return {@code null}.
+   */
+  protected final WorkflowContext getContext() {
```
annotate as nullable

albertshau replied:
actually, we don't annotate the other ones as nullable.  Because we don't want people to call this when its not at runtime?

sagarkapare replied:
my bad.. that comment is wrong. It should always return `WorkflowContext`. The comment should just be `Return an instance of the {@link WorkflowContext}`. updating the comment.

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java:
```diff
@@ -173,6 +181,30 @@ protected void startUp() throws Exception {
+    try {
+      workflow.initialize(basicWorkflowContext);
+    } catch (Throwable t) {
+      LOG.error("Failed to execute the initialize method of the Workflow.", t);
```
do we have the workflow name anywhere? Would be nice for the message to say which one failed.

Also nit: "Failed to initialize workflow xyz" would be a more concise message

sagarkapare replied:
Yes. We do have access to the `ProgramRunId` instance. Using that in the error message.

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java:
```diff
@@ -222,6 +254,23 @@ public void resume() throws Exception {
   protected void shutDown() throws Exception {
     httpService.stopAndWait();
     deleteLocalDatasets();
+    destroyWorkflow();
```
can deleteLocalDatasets() throw exceptions? if so we need to catch and make sure destroyWorkflow is run

sagarkapare replied:
`deleteLocalDatasets()` does not throw exception.

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java:
```diff
@@ -222,6 +254,23 @@ public void resume() throws Exception {
+      workflow.destroy();
+      transactionContext.finish();
+    } catch (Throwable t) {
+      LOG.error("Failed to execute the destroy method of the Workflow.", t);
```
similar comment, would be nice to have the workflow name and namespace in the error message

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/AbstractWorkflow.java:
```diff
@@ -189,4 +190,21 @@ protected final void createLocalDataset(String datasetName, Class<? extends Data
+
+  @Override
+  public void destroy() {
+    // Nothing to do.
```
nit: "// Do nothing" is a little more accurate

sagarkapare replied:
fixed.

albertshau commented:
small comments about error messages, otherwise :+1:

sagarkapare commented:
Thanks @albertshau for the review. Addressed comments and added one more check in the unit test case to make sure that Workflow runs are getting completed even the destroy throws an exception.

New build is running here - http://builds.cask.co/browse/CDAP-DUT3732-2. Will merge once the build passes.

sagarkapare commented:
Build is green - http://builds.cask.co/browse/CDAP-DUT3732-3. Merging.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/Workflow.java:
```diff
@@ -15,13 +15,15 @@
  * Defines an interface for the Workflow.
  */
-public interface Workflow {
+public interface Workflow extends ProgramLifecycle<WorkflowContext> {
```
This change is not backward compatible. You can't expect everyone is extending from AbstractWorkflow

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java:
```diff
@@ -173,6 +181,30 @@ protected void startUp() throws Exception {
+    transactionContext.start();
+
+    try {
+      workflow.initialize(basicWorkflowContext);
```
Workflow implementation of ProgramLifecycle should be optional, hence you need an instance of check before calling lifecycle methods. Same goes for destroy
